xcm/pallet-xcm/src/lib.rs

```diff
@@ -635,87 +1044,133 @@ pub mod pallet {
 			response: Response,
 			max_weight: Weight,
 		) -> Weight {
-			if let Some(QueryStatus::Pending { responder, maybe_notify, .. }) =
-				Queries::<T>::get(query_id)
-			{
-				if let Ok(responder) = MultiLocation::try_from(responder) {
-					if origin == &responder {
-						return match maybe_notify {
-							Some((pallet_index, call_index)) => {
-								// This is a bit horrible, but we happen to know that the `Call` will
-								// be built by `(pallet_index: u8, call_index: u8, QueryId, Response)`.
-								// So we just encode that and then re-encode to a real Call.
-								let bare = (pallet_index, call_index, query_id, response);
-								if let Ok(call) = bare.using_encoded(|mut bytes| {
-									<T as Config>::Call::decode(&mut bytes)
-								}) {
-									Queries::<T>::remove(query_id);
-									let weight = call.get_dispatch_info().weight;
-									if weight > max_weight {
-										let e = Event::NotifyOverweight(
-											query_id,
-											pallet_index,
-											call_index,
-											weight,
-											max_weight,
-										);
-										Self::deposit_event(e);
-										return 0
-									}
-									let dispatch_origin = Origin::Response(origin.clone()).into();
-									match call.dispatch(dispatch_origin) {
-										Ok(post_info) => {
-											let e =
-												Event::Notified(query_id, pallet_index, call_index);
-											Self::deposit_event(e);
-											post_info.actual_weight
-										},
-										Err(error_and_info) => {
-											let e = Event::NotifyDispatchError(
-												query_id,
-												pallet_index,
-												call_index,
-											);
-											Self::deposit_event(e);
-											// Not much to do with the result as it is. It's up to the parachain to ensure that the
-											// message makes sense.
-											error_and_info.post_info.actual_weight
-										},
-									}
-									.unwrap_or(weight)
-								} else {
-									let e = Event::NotifyDecodeFailed(
+			match (response, Queries::<T>::get(query_id)) {
```

Comment:
in terms of weighing this function, it would be helpful to break down this large match statement into sub-functions, which would then allow us to weight the individual logical paths in a more certain and controllable way. If you want, this refactor can happen in #3735